File: Python/BinarySearch/입국심사.py
def solution(n, times):
    answer = 0
    
    # left - the most effective case, minimum time
    # right - the most ineffective case, maximum time
    left = 1 # case; only 1 waiting person, only 1 minute needed
    right = max(times)*n # case; the slowest man deal to all n people
    
    while left <= right:
        mid = (left + right)//2
        
        # people - total number of people passed during mid-minutes
        people=0
        for time in times:
            people += mid//time
            if people >= n:
                break
        
        #if the total number of people passed is equl to or bigger than n, go to left-tree
        if people >= n:
            right = mid - 1
            answer = mid
        
        #if the total number of people passed is smaller than n, go to right-tree
        if people < n:
            left = mid + 1
            
    return answer

File: C++/BinarySearch/입국심사.cpp
#include <string>
#include <vector>
#include <algorithm>
using namespace std;

long long solution(int n, vector<int> times) {
    long long answer = 0;
    long long left, mid, right, people;
    
    // left: 가능한 최소 시간, right: 가능한 최대 시간
    left = 1;
    right = *max_element(times.begin(), times.end()) * (long long)n;
    
    while (left <= right) {
        people = 0;
        mid = (left + right) / 2;
        
        // people: mid 시간동안 심사할 수 있는 모든 사람 수
        for (int i = 0; i < times.size(); i++) {
            people += mid / times[i];
        }

        // people이 n명 이상을 충족한 경우 right idx 변경 & answer 갱신
        if (n <= people) {
            answer = mid;
            right = mid - 1;
        }

        // people이 n명보다 작다면 조건 만족하지 않으므로 left idx 변경
        else {
            left = mid + 1;
        }
    }

    return answer;
}